pkg/kubelet/status/state/checkpoint.go

```diff
@@ -18,48 +18,62 @@ package state
+var _ checkpointmanager.Checkpoint = &Checkpoint{}
 
-// PodResourceAllocationCheckpoint is used to store resources allocated to a pod in checkpoint
-type PodResourceAllocationCheckpoint struct {
+type PodResourceAllocationInfo struct {
 	AllocationEntries   map[string]map[string]v1.ResourceList `json:"allocationEntries,omitempty"`
 	ResizeStatusEntries map[string]v1.PodResizeStatus         `json:"resizeStatusEntries,omitempty"`
-	Checksum            checksum.Checksum                     `json:"checksum"`
 }
 
-// NewPodResourceAllocationCheckpoint returns an instance of Checkpoint
-func NewPodResourceAllocationCheckpoint() *PodResourceAllocationCheckpoint {
-	//lint:ignore unexported-type-in-api user-facing error message
-	return &PodResourceAllocationCheckpoint{
-		AllocationEntries:   make(map[string]map[string]v1.ResourceList),
-		ResizeStatusEntries: make(map[string]v1.PodResizeStatus),
+// Checkpoint represents a structure to store pod resource allocation checkpoint data
+type Checkpoint struct {
+	// Data is a JSON serialized checkpoint data
+	Data string `json:"data"`
+	// Checksum is a checksum of Data
+	Checksum checksum.Checksum `json:"checksum"`
+}
+
+// NewCheckpoint creates a new checkpoint from a list of claim info states
+func NewCheckpoint(data *PodResourceAllocationInfo) (*Checkpoint, error) {
+
+	praData, err := json.Marshal(data)
+	if err != nil {
+		return nil, err
+	}
+
+	cp := &Checkpoint{
+		Data: string(praData),
 	}
+	cp.Checksum = checksum.New(cp.Data)
+	return cp, nil
+}
+
+func (cp *Checkpoint) MarshalCheckpoint() ([]byte, error) {
+	return json.Marshal(cp)
 }
 
-// MarshalCheckpoint returns marshalled checkpoint
-func (prc *PodResourceAllocationCheckpoint) MarshalCheckpoint() ([]byte, error) {
-	// make sure checksum wasn't set before so it doesn't affect output checksum
-	prc.Checksum = 0
-	prc.Checksum = checksum.New(prc)
-	return json.Marshal(*prc)
+// UnmarshalCheckpoint unmarshals checkpoint from JSON
+func (cp *Checkpoint) UnmarshalCheckpoint(blob []byte) error {
+	return json.Unmarshal(blob, cp)
 }
 
-// UnmarshalCheckpoint tries to unmarshal passed bytes to checkpoint
-func (prc *PodResourceAllocationCheckpoint) UnmarshalCheckpoint(blob []byte) error {
-	return json.Unmarshal(blob, prc)
+// VerifyChecksum verifies that current checksum
+// of checkpointed Data is valid
+func (cp *Checkpoint) VerifyChecksum() error {
+	return cp.Checksum.Verify(cp.Data)
 }
 
-// VerifyChecksum verifies that current checksum of checkpoint is valid
-func (prc *PodResourceAllocationCheckpoint) VerifyChecksum() error {
-	ck := prc.Checksum
-	prc.Checksum = 0
-	err := ck.Verify(prc)
-	prc.Checksum = ck
-	return err
+// GetPodResourceAllocationInfo returns Pod Resource Allocation info states from checkpoint
+func (cp *Checkpoint) GetPodResourceAllocationInfo() (*PodResourceAllocationInfo, error) {
+	var data PodResourceAllocationInfo
+	if err := json.Unmarshal([]byte(cp.Data), &data); err != nil {
```

Comment:
This un-marshals data without verifying checksum. Should cp.UnmarshalCheckpoint be called here instead?

Reply:
@bart0sh No, it shouldn't in my opinion.
Verifying checksum is necessary immediately after reading the checkpoint from the file, but this function is invoked at the situation that the checkpoint already is read into struct `Checkpoint`, which is in memory.
> file -> ① -> Checkpoint -> ② -> stateCheckpoint

It's necessary for procedure ① to verify checksum, but not for procedure ②.
And function `GetPodResourceAllocationInfo` is invoked at procedure ②. 
```golang
func (sc *stateCheckpoint) restoreState() error {
	...
        // procudure ① 
	if err = sc.checkpointManager.GetCheckpoint(sc.checkpointName, checkpoint); err != nil {
		if err == errors.ErrCheckpointNotFound {
			return sc.storeState()
		}
		return err
	}
        // procudure ②
	praInfo, err := checkpoint.GetPodResourceAllocationInfo()
	if err != nil {
		return fmt.Errorf("failed to get pod resource allocation info: %w", err)
	}
	...
}
```